### DISCUSSION TITLE
CPU optimizations when running virtualized routers

### DISCUSSION THREAD

**Discussion by pedroaston** [srl-labs/containerlab#1805] [Q&A] [answered]
## Context
I'm trying to scale a topology so I can emulate an entire ISP network (Enterprise CPEs, Core SR/MPLS, DC). Since routers in SR/MPLS core tend to be virtualized routers and I still need some of them, I used KSM to optimize memory. The result was that I still had a lot of spare memory, but CPU workloads continued to scale linearly.

## Question
Is there a way to optimize CPU usage without requiring 2 vCPUs per VM router?

## Resource utilization
Scenario:
6 VM-Routers with xrv9k:7.11.1 image (Requisites >> 14G RAM and 2 vCPUs)
- 4 PEs and 2 Ps
![image](https://github.com/srl-labs/containerlab/assets/36762762/53dc2f92-6a2c-47a4-87d3-e82e06436d0c)

Thanks in advance

**Comment by hellt** [comment, marked as answer]
Hi, not really, no. I think the CPU scaling would be solely in hands of the NOS you're working with.

The only possible way is to check the NOS documentation for some CPU tuning, but I doubt there is anything that can substantially reduce the CPU load.

**Reply by pedroaston** [reply]
Ok thought so, thanks for the feedback.
